Pass user-provided NUTS kwargs to Numpyro

jhrcook opened the pull request:
<!-- !! Thank your for opening a PR !! -->

**What is this PR about?**

There are currently defaults for a few keyword arguments for the Numpyro NUTS initialization method that prevent the PyMC user for passing certain keyword arguments. This PR refactors the code to update the user-provided keyword arguments dictionary with the defaults if they weren't included.

Addresses issue #6020 

**Checklist**
+ [x] Explain important implementation details 👆
+ [x] Make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).
+ [x] Link relevant issues (preferably in [nice commit messages](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html))
+ [x] Are the changes covered by tests and docstrings?
+ [x] Fill out the short summary sections 👇

## Major / Breaking Changes

## Bugfixes / New features
- Enable the user to pass any keyword arguments to the Numpyro NUTS initialization.

## Docs / Maintenance
- Updated docstring of `sample_numpyro_nuts` to include information for the `initvals` argument
- fixed some incorrect typehints of arguments in the `sample_numpyro_nuts` definition
- added a return typehint to `sample_numpyro_nuts`

jhrcook committed:
refactor: update user supplied NUTS kwargs with defualts for Numpyro sampler

jhrcook committed:
style: fix typehint in `sample_numpyro_nuts`

jhrcook committed:
doc: add missing `initvals` to docstring of `sample_numpyro_nuts`

jhrcook committed:
style: return typehint for `sample_numpyro_nuts`

jhrcook committed:
refactor: change `var_names` typehint from `Iterable` to `Sequence`

twiecki commented:
@jhrcook Can you add a test?

jhrcook committed:
Merge branch 'pymc-devs:main' into numpyro-kwargs

jhrcook committed:
test: updating NUTS kwargs for Numpyro sampler function

jhrcook commented:
@twiecki  I've added tests for the keyword-argument updating function and another set of tests that run a short MCMC with some different NUTS keyword arguments, but I have limited ability to actually determine if `numpyro.NUTS` actually *used* the arguments. The only one I could think of was to set the `step_size` and confirm that it was in the result `trace.sample_stats`. I could try monkey-patching into the NUTS initializer method, but it could be tricky because the `sample_numpyro_nuts` imports `NUTS` internally. Let me know if you'd like me to try, though.

ricardoV94 commented on pymc/tests/test_sampling_jax.py:
```diff
@@ -270,3 +275,47 @@ def test_seeding(chains, random_seed, sampler):
+        {"adapt_step_size": False, "adapt_mass_matrix": True, "dense_mass": True},
+        {"adapt_step_size": False, "step_size": 0.13},
+    ],
+)
```
Let's try to monkey patch? If it's not possible I think we can survive without explicit tests. This one seems too inneficient.

jhrcook replied:
I'll give it shot. I agree that the current method I used is inefficient and isn't really testing the desired behavior.

jhrcook committed:
test: use monkeypatch to determine if custum NUTS kwargs are used

jhrcook commented:
I used a monkeypatch to intercept the `NUTS` object in the initialization of the Numpyro `MCMC` object when called in `pymc.sampling_jax.sample_numpyro_jax()`. I just made a `MockMCMC` that checks some of the properties of its `NUTS` sampler and asserts they are what I passed to `sample_numpyro_jax(..., nuts_kwargs={...})`.

I confirmed that the monkeypatch was intercepting the `MCMC.__init__()` by purposefully making the tests fail with incorrect assertions.

I have limited experience with using monkeypatches in tests, so let me know if you think there would be a better construction/format.

ricardoV94 commented:
When using monkey patching there's a helper method, something like `assert_called_once_with(...)` where you can directly test your monkey patched object was called once with specific arguments and kwargs.

jhrcook committed:
test: replace monkey patch with simpler mocking method

jhrcook commented:
Thank you @ricardoV94 for the help. I've replaced the monkey patching system with a rather simple (too simple?) mocking system. I used mocking to allow me to peer into when the `numpyro.infer.MCMC` object is made and then inspect the NUTS sampler it was passed. This puts all of the assertions in a single function and removed the need to make a new mock class.

ricardoV94 commented:
> Thank you @ricardoV94 for the help. I've replaced the monkey patching system with a rather simple (too simple?) mocking system. I used mocking to allow me to peer into when the `numpyro.infer.MCMC` object is made and then inspect the NUTS sampler it was passed. This puts all of the assertions in a single function and removed the need to make a new mock class.

Looks perfect. The simpler the better!

twiecki commented:
Thanks @jhrcook!
